src/measure_theory/function/jacobian.lean

```diff
@@ -0,0 +1,1274 @@
+      filter_upwards [(tendsto_order.1 L).2 _ hy.2],
+      assume k hk,
+      exact (aM k).2 y ⟨hy.1, hk⟩ },
+    apply le_of_tendsto_of_tendsto L1 L2 I },
+  -- choose a dense sequence `d p`
+  rcases topological_space.exists_dense_seq E with ⟨d, hd⟩,
+  -- split `M n z` into subsets `K n z p` of small diameters by intersecting with the ball
+  -- `closed_ball (d p) (u n / 3)`.
+  let K : ℕ → T → ℕ → set E := λ n z p, closure (M n z) ∩ closed_ball (d p) (u n / 3),
+  -- on the sets `K n z p`, the map `f` is well approximated by `f' z` by design.
+  have K_approx : ∀ n (z : T) p, approximates_linear_on f (f' z) (s ∩ K n z p) (r (f' z)),
+  { assume n z p x hx y hy,
+    have yM : y ∈ M n z := closure_M_subset _ _ ⟨hy.1, hy.2.1⟩,
+    refine yM.2 _ ⟨hx.1, _⟩,
+    calc dist x y ≤ dist x (d p) + dist y (d p) : dist_triangle_right _ _ _
+    ... ≤ u n / 3 + u n / 3 : add_le_add hx.2.2 hy.2.2
+    ... < u n : by linarith [u_pos n] },
+  -- the sets `K n z p` are also closed, again by design.
+  have K_closed : ∀ n (z : T) p, is_closed (K n z p) :=
+    λ n z p, is_closed_closure.inter is_closed_ball,
+  -- reindex the sets `K n z p`, to let them only depend on an integer parameter `q`.
+  obtain ⟨F, hF⟩ : ∃ F : ℕ → ℕ × T × ℕ, function.surjective F,
+  { haveI : encodable T := T_count.to_encodable,
+    haveI : nonempty T,
+    { unfreezingI { rcases eq_empty_or_nonempty T with rfl|hT },
+      { rcases hs with ⟨x, xs⟩,
+        rcases s_subset x xs with ⟨n, z, hnz⟩,
+        exact false.elim z.2 },
+      { exact (nonempty_coe_sort _).2 hT } },
+    inhabit (ℕ × T × ℕ),
+    exact ⟨_, encodable.surjective_decode_iget _⟩ },
+  -- these sets `t q = K n z p` will do
+  refine ⟨λ q, K (F q).1 (F q).2.1 (F q).2.2, λ q, f' (F q).2.1, λ n, K_closed _ _ _, λ x xs, _,
+    λ q, K_approx _ _ _, λ h's q, ⟨(F q).2.1, (F q).2.1.1.2, rfl⟩⟩,
+  -- the only fact that needs further checking is that they cover `s`.
+  -- we already know that any point `x ∈ s` belongs to a set `M n z`.
+  obtain ⟨n, z, hnz⟩ : ∃ (n : ℕ) (z : T), x ∈ M n z := s_subset x xs,
+  -- by density, it also belongs to a ball `closed_ball (d p) (u n / 3)`.
+  obtain ⟨p, hp⟩ : ∃ (p : ℕ), x ∈ closed_ball (d p) (u n / 3),
+  { have : set.nonempty (ball x (u n / 3)),
+    { simp only [nonempty_ball], linarith [u_pos n] },
+    obtain ⟨p, hp⟩ : ∃ (p : ℕ), d p ∈ ball x (u n / 3) := hd.exists_mem_open is_open_ball this,
+    exact ⟨p, (mem_ball'.1 hp).le⟩ },
+  -- choose `q` for which `t q = K n z p`.
+  obtain ⟨q, hq⟩ : ∃ q, F q = (n, z, p) := hF _,
+  -- then `x` belongs to `t q`.
+  apply mem_Union.2 ⟨q, _⟩,
+  simp only [hq, subset_closure hnz, hp, mem_inter_eq, and_self],
+end
+
+variables [measurable_space E] [borel_space E] (μ : measure E) [is_add_haar_measure μ]
+
+/-- Assume that a function `f` has a derivative at every point of a set `s`. Then one may
+partition `s` into countably many disjoint relatively measurable sets `t n` on which `f` is well
```

Comment:
what do you mean by "relatively measurable sets"?

Reply:
I have explained that in the docstring.